Assets/BossRoom/Scripts/Server/Game/AIState/IdleAIState.cs

```diff
@@ -33,11 +34,18 @@ protected void DetectFoes()
             float detectionRangeSqr = detectionRange * detectionRange;
             Vector3 position = m_Brain.GetMyServerCharacter().transform.position;
 
-            foreach (ServerCharacter character in ServerCharacter.GetAllActiveServerCharacters())
+            foreach (var spawnedObject in NetworkSpawnManager.SpawnedObjectsList)
             {
-                if (m_Brain.IsAppropriateFoe(character) && (character.transform.position - position).sqrMagnitude <= detectionRangeSqr)
+                if (!spawnedObject) { continue; } // must have been Destroy()ed very recently
+                if ((spawnedObject.transform.position - position).sqrMagnitude <= detectionRangeSqr)
                 {
-                    m_Brain.Hate(character);
+                    // they're within range... are they an appropriate foe?
+                    ServerCharacter serverChar = spawnedObject.GetComponent<ServerCharacter>();
```

Comment:
So just with a quick deep profile, this change allocates ~50 KB of memory per frame as it's called ~300 times. Let's keep the old approach with the static collection of `ServerCharacter`s and refactor the places in code to use this to be consistent.

Reply:
sounds good

Reply:
funnily enough I don't see any other uses that need to be switched to use this list. The other uses were scenarios where we needed to iterate over players (not just all characters), so after refactoring, they use `NetworkManager.Singleton.ConnectedClients` instead. And since that list is much shorter than the list of all characters, that's probably the more efficient approach for them.

So in the end this PR is just some comment clean up!

Reply:
Why is it iterating over all characters? Why not just players?

Reply:
Hm... actually yeah, it can be. Back when we started, I wanted to support the notion of "pets" (e.g. a "sentient projectile" that could attract enemy attention), so enemies search for "foes" rather than just players. But none of that has ended up being relevant. 

Unless you object I'll go ahead and change it to a player-list iteration. That obsoletes the list of ServerCharacters, so I'll just remove that.

Reply:
thank you

Reply:
ok

Reply:
We want to eliminate all possible GC allocations wherever possible. There may be a solution here that fits both needs.

How about enforcing composition here? We attach a `PlayerServerCharacter` component (that requires a `ServerCharacter`) on our player prefab. This defines a publicly static collection like before. This component specifically adds and removes players to this collection on OnEnable()/OnDisable() respectively. That way you'd be getting a reference to all & only player `ServerCharacter`s without any `GetComponent<>`s.

Reply:
Okay, I've added the `PlayerServerCharacter` component to the players' prefab and refactored `IdleAIState` (and a few other places) to use this list.